core/src/main/scala/org/http4s/MessageFailure.scala

```diff
@@ -80,7 +80,8 @@ object ParseResult {
 
   private[http4s] def fromParser[A](parser: Parser0[A], errorMessage: => String)(
       s: String): ParseResult[A] =
-    parser.parseAll(s).leftMap(e => ParseFailure(errorMessage, e.toString))
+    try parser.parseAll(s).leftMap(e => ParseFailure(errorMessage, e.toString))
+    catch { case p: ParseFailure => p.asLeft[A] }
```

Comment:
I'd say this is a bug if it happens, and we might want to screech about it in the logs.  But the defensive programming is good.